rslib/src/notetype/mod.rs

```diff
@@ -504,14 +496,33 @@ impl Notetype {
         }
         self.config.reqs = reqs;
         if !skip_checks {
-            self.ensure_template_fronts_unique()?;
-            self.ensure_valid_parsed_templates(&parsed_templates)?;
-            self.ensure_cloze_if_cloze_notetype(&parsed_templates)?;
+            self.check_templates(parsed_templates)?;
         }
 
         Ok(())
     }
 
+    fn ensure_unempty_values(&mut self) -> Result<(), InvalidInputError> {
+        let bad_chars = |c| c == '"';
+        if self.name.contains(bad_chars) {
+            self.name = self.name.replace(bad_chars, "");
+        }
+        ensure_whatever!(!self.name.is_empty(), "Empty note type name");
+        ensure_whatever!(!self.fields.is_empty(), "1 field required");
+        ensure_whatever!(!self.templates.is_empty(), "1 template required");
```

Comment:
`whatever` seems to not work well with the coercion into a containing error, so if this routine had to handle some other type of error as well, I don't think we could use this. The following code fails for example:

```rust
fn map_example() -> Result<(), FutureAnkiError> {
    // legacy style
    let ctype = CardType::try_from(7)
        .map_err(|_| AnkiError::invalid_input("invalid card type"))?;
    // whatever
    let ctype = whatever!(CardType::try_from(7), "invalid card type");
Ok(())
}
```

Maybe we'd be better off with a helper routine and avoiding `whatever`?

Reply:
Not sure I understand your example. Wouldn't InvalidInput always be a top-level error anyway?

Reply:
By that, I guess you mean defining it as a variant of FutureAnkiError directly? What if we want more than one variant with string errors there?

Reply:
Yes. I can't think of another use case for string errors apart from unexpectedly invalid input, but should one arise in the future, we could still add a helper routine for that and keep using `whatever` for invalid input, couldn't we?
This seems to require the least amount of work for now and in the long run.

Reply:
I can't think of another of our existing variants that would benefit from string errors at the moment. Presumably it would also rule out the usage of .(with_)whatever_context()? Not necessarily a bad thing, as I guess that would complicate later determining the class of error if I've understood things correctly (we'd need to hunt for it in the source error?)

The examples above focused on throwing an error based on a condition - perhaps it would be worth also experimenting a bit with how we'd deal with existing errors that need to be mapped to a different error with context added, since that's also common?

Reply:
> The examples above focused on throwing an error based on a condition - perhaps it would be worth also experimenting a bit with how we'd deal with existing errors that need to be mapped to a different error with context added, since that's also common?

I don't know what you mean exactly. Is this still about InvalidInput? Do you have an example?

Reply:
Sorry, not about InvalidInput, just about using snafu in general. For example, the way we include a filename in data_for_file() and atomic_rename(), and things like 

```rust
        let decks = self
            .get_schema11_decks()
            .map_err(|e| AnkiError::JsonError(format!("decoding decks: {}", e)))?;
```

Reply:
> The examples above focused on throwing an error based on a condition - perhaps it would be worth also experimenting a bit with how we'd deal with existing errors that need to be mapped to a different error with context added, since that's also common?

I've added `FileIoError` now.